### PR TITLE
[Do not merge-need to fix reference]Adding NotificationHub Scripting

### PR DESCRIPTION
**Pull request by pragnagopa**

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@pragnagopa__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (pgopa). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: sample/host.json [hunk 1]
```diff
@@ -1,3 +1,4 @@
 {
-  "id": "5a709861cab44e68bfed5d2c2fe7fc0c"
+  "id": "5a709861cab44e68bfed5d2c2fe7fc0c",
+  "functions": [ "NotificationHub-Node" ]
 }
```
Comment (brettsam) [inline sample/host.json]: make sure to back this out

File: src/WebJobs.Script/app.config [hunk 2]
```diff
@@ -42,5 +42,6 @@
 	</runtime>
 <startup><supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6" /></startup>
 <appSettings>
+<add key="Microsoft.Azure.NotificationHubs.ConnectionString" value="Endpoint=sb://[your namespace].notificationhubs.windows.net;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=[your secret]" />
 </appSettings>
```
Comment (brettsam) [inline src/WebJobs.Script/app.config]: don't think this is needed

File: src/WebJobs.Script.Host/App.config [hunk 3]
```diff
@@ -40,4 +40,9 @@
       </dependentAssembly>
     </assemblyBinding>
   </runtime>
-<startup><supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6" /></startup></configuration>
+<startup><supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6" /></startup>
+<appSettings>
+<!-- NotificationHubs specific app setings for messaging connections -->
+<add key="Microsoft.Azure.NotificationHubs.ConnectionString" value="Endpoint=sb://[your namespace].notificationhubs.windows.net;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=[your secret]" />
+</appSettings>
```
Comment (brettsam) [inline src/WebJobs.Script.Host/App.config]: not needed

File: src/WebJobs.Script/Binding/NotificationHubBinding.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                TagExpression
+            };
+
+            ConstructorInfo constructor = attributeType.GetConstructor(System.Type.EmptyTypes);
+
+            return new CustomAttributeBuilder(constructor, new object[] { }, props, propValues);
+        }
+
+        public override async Task BindAsync(BindingContext context)
+        {
```
Comment (brettsam) [inline src/WebJobs.Script/Binding/NotificationHubBinding.cs]: add comment that this only supports json strings.

File: src/WebJobs.Script/Description/NotificationHubBindingMetadata.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            {
+                throw new ArgumentNullException("configBuilder");
+            }
+
+            NotificationHubConfiguration config = new NotificationHubConfiguration();
+            if (!string.IsNullOrEmpty(ConnectionString) && !string.IsNullOrEmpty(HubName))
+            {
+                config.ConnectionString = ConnectionString;
+                config.HubName = HubName;
+            }
```
Comment (brettsam) [inline src/WebJobs.Script/Description/NotificationHubBindingMetadata.cs]: change this to two separate if blocks -- it's possible that ConnectionString is set in the config yet HubName is not (or vice versa)

File: test/WebJobs.Script.Tests/CSharpEndToEndTests.cs [hunk 6]
```diff
@@ -32,6 +32,25 @@ public CSharpEndToEndTests(TestFixture fixture) : base(fixture)
...
 
         [Fact]
+        public async Task NotificationHub()
+        {
+            await NotificationHubTest();
+        }
+
+        [Fact]
+        public async Task NotificationHub_Out_Notification()
+        {
```
Comment (brettsam) [inline test/WebJobs.Script.Tests/CSharpEndToEndTests.cs]: I think you can just do:

TouchProjectJson()
NotificationHubTest()

I think the other stuff is just duplicated from your NH test method.

File: test/WebJobs.Script.Tests/EndToEndTestsBase.cs [hunk 7]
```diff
@@ -69,6 +69,18 @@ public EndToEndTestsBase(TTestFixture fixture)
             Assert.Equal(doc.Id, id);
         }
 
+        protected async Task NotificationHubTest()
+        {
```
Comment (brettsam) [inline test/WebJobs.Script.Tests/EndToEndTestsBase.cs]: comment here that there's no actual validation done b/c it's hard (impossible?) to verify from within this project. Instead, we're making sure the function is called and exits without throwing.

File: sample/NotificationHub-CSharp/function.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+  "bindings": [
+    {
+      "type": "queueTrigger",
+      "direction": "in",
+      "queueName": "samples-notificationhub-csharp"
+    },
+    {
+      "type": "notificationhub",
+      "direction": "out",
+      "name": "messageProperties"
```
Comment (mathewc) [inline sample/NotificationHub-CSharp/function.json]: Wondering if "messageProperties" is a good name - how about just "message"?

File: sample/NotificationHub-CSharp/run.csx [hunk 9]
```diff
@@ -0,0 +1,5 @@
+﻿using System;
+public static void Run(string input, out string messageProperties)
+{
+    messageProperties = "{\"message\":\"Hello from c#! \",\"location\":\"Redmond\"}";
```
Comment (mathewc) [inline sample/NotificationHub-CSharp/run.csx]: capitalize C# :)

File: sample/host.json (hunk 1 above)
Comment (mathewc) [inline sample/host.json]: My new feature is coming in handy :) I've been using it alot too :)

File: src/WebJobs.Script.Host/App.config (hunk 3 above)
Comment (mathewc) [inline src/WebJobs.Script.Host/App.config]: yeah please remove

File: src/WebJobs.Script/WebJobs.Script.csproj [hunk 10]
```diff
@@ -52,6 +52,14 @@
       <HintPath>..\..\packages\Edge.js.4.0.0\lib\EdgeJs.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="FSharp.Core, Version=4.3.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a, processorArchitecture=MSIL">
```
Comment (mathewc) [inline src/WebJobs.Script/WebJobs.Script.csproj]: Where are these two F# references coming from? They used to come from us importing the old WebHooks extension, but we no longer do so I don't see why they're here?

File: src/WebJobs.Script/WebJobs.Script.csproj [hunk 11]
```diff
@@ -64,24 +72,36 @@
...
+      <HintPath>..\..\packages\Microsoft.Azure.NotificationHubs.1.0.3\lib\net45-full\Microsoft.Azure.NotificationHubs.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
     <Reference Include="Microsoft.Azure.WebJobs, Version=1.1.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
       <HintPath>..\..\packages\Microsoft.Azure.WebJobs.Core.1.1.2-alpha-10267\lib\net45\Microsoft.Azure.WebJobs.dll</HintPath>
       <Private>True</Private>
     </Reference>
     <Reference Include="Microsoft.Azure.WebJobs.Extensions, Version=1.0.2.0, Culture=neutral, processorArchitecture=MSIL">
-      <HintPath>..\..\packages\Microsoft.Azure.WebJobs.Extensions.1.0.2-alpha-10252\lib\net45\Microsoft.Azure.WebJobs.Extensions.dll</HintPath>
+      <HintPath>..\..\packages\Microsoft.Azure.WebJobs.Extensions.1.0.2-alpha\lib\net45\Microsoft.Azure.WebJobs.Extensions.dll</HintPath>
```
Comment (mathewc) [inline src/WebJobs.Script/WebJobs.Script.csproj]: Ah, these are local package refs that you have to fix up

File: test/WebJobs.Script.Tests/CSharpEndToEndTests.cs (hunk 6 above)
Comment (mathewc) [inline test/WebJobs.Script.Tests/CSharpEndToEndTests.cs]: Regarding the TouchProjectJson - why is that needed? If a function includes a project.json, it should be discovered on startup - I.e., I don't see why you need ScriptHostManager + file watching for these tests.

File: test/WebJobs.Script.Tests/EndToEndTestsBase.cs (hunk 7 above)
Comment (mathewc) [inline test/WebJobs.Script.Tests/EndToEndTestsBase.cs]: We could have the message sent to you Brett, then you could manually increment a value in table store, and she could poll for that in the test. You'd have to remain responsive 24/7 though so the tests run reliably.

File: test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOut/function.json [hunk 12]
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "bindings": [
+    {
+      "type": "manualTrigger",
+      "direction": "in"
+    },
+    {
+      "type": "notificationhub",
+      "name": "messageProperties",
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOut/function.json]: "message"?

File: test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOutNotification/function.json [hunk 13]
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "bindings": [
+    {
+      "type": "manualTrigger",
+      "direction": "in"
+    },
+    {
+      "type": "notificationhub",
+      "name": "notification",
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOutNotification/function.json]: maybe use consistent names for these in all your samples/tests?

File: test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOut/run.csx [hunk 14]
```diff
@@ -0,0 +1,5 @@
+﻿using System;
+public static void Run(string input, out string messageProperties)
+{
+    messageProperties = "{\"message\":\"Hello from c#! \",\"location\":\"Redmond\"}";
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOut/run.csx]: C#

File: test/WebJobs.Script.Tests/EndToEndTestsBase.cs [hunk 15]
```diff
@@ -69,6 +69,18 @@ public EndToEndTestsBase(TTestFixture fixture)
...
+            // NotificationHub tests need the following environment vars:
+            // "AzureWebJobsNotificationHubsConnectionString" -- the connection string for NotificationHubs
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/EndToEndTestsBase.cs]: Please send a not to team on setting this up correctly when this gets in, so we can update our machines to run the tests successfully.

File: test/WebJobs.Script.Tests/TestScripts/Node/NotificationHubOut/function.json [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+  "bindings": [
+    {
+      "type": "manualTrigger",
+      "direction": "in"
+    },
+    {
+      "type": "notficationhub",
+      "name": "notification",
+      "direction": "out",
+      "tagexpression": "foo||bar"
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Node/NotificationHubOut/function.json]: What does this do? Allow NH subscription filtering on tags? Do you support binding parameters on this - e.g. a queue message comes in that has one or more tags defined in it, which you can auto bind?

File: test/WebJobs.Script.Tests/TestScripts/Node/NotificationHubOut/index.js [hunk 17]
```diff
@@ -0,0 +1,10 @@
+﻿module.exports = function (context, input) {
+    context.log('Node.js triggered function via NotificationHub called with input ' + input);
+    //TODO
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Node/NotificationHubOut/index.js]: Remove TODO

File: test/WebJobs.Script.Tests/CSharpEndToEndTests.cs (hunk 6 above)
Comment (brettsam) [inline test/WebJobs.Script.Tests/CSharpEndToEndTests.cs]: I talked with @fabiocav while I was writing this. It's by-design that nuget does not do a restore on startup and instead only reacts when project.json is changed. If I remember the conversation correctly -- in practice, the host will always be running when a project.json file appears or is changed. So our e2e tests need to simulate that.

File: test/WebJobs.Script.Tests/TestScripts/Node/NotificationHubOut/index.js [hunk 18]
```diff
@@ -0,0 +1,10 @@
+﻿module.exports = function (context, input) {
+    context.log('Node.js triggered function via NotificationHub called with input ' + input);
```
Comment (mathewc) [inline test/WebJobs.Script.Tests/TestScripts/Node/NotificationHubOut/index.js]: Not via - that seems to imply that NH triggered it

File: test/WebJobs.Script.Tests/CSharpEndToEndTests.cs (hunk 6 above)
Comment (mathewc) [inline test/WebJobs.Script.Tests/CSharpEndToEndTests.cs]: I'm wondering about the scenario where someone git pushes their functions with a project.json, to a site where the host is not running (e.g. new site creation). When it starts up, i'd expect stuff to just work. @fabiocav?

Comment (brettsam) [inline test/WebJobs.Script.Tests/CSharpEndToEndTests.cs]: @mattchenderson and I were just talking about this. He did git deploy and it didn't restore his nugets. I think he's working with someone on it.

Comment (mathewc) [inline test/WebJobs.Script.Tests/CSharpEndToEndTests.cs]: Yeah, I'd expect that scenario to work. When it does, I think we should move back to simplifying these tests to use ScriptHost directly - they're more complicated than they need to be I think.

File: sample/NotificationHub-CSharp/function.json (hunk 8 above)
Comment (pragnagopa) [inline sample/NotificationHub-CSharp/function.json]: messageProperties - is JSON string with key value pairs that will be used to build templateNotification

File: test/WebJobs.Script.Tests/EndToEndTestsBase.cs (hunk 7 above)
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/EndToEndTestsBase.cs]: For now, just verifying the call succeeds. Will add more verification later. There might be a way to call NH api to check if sendAsync was called.

File: test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOut/function.json (hunk 12 above)
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOut/function.json]: same here. messageProperties is key value pairs that will be used to build templateNotification

File: test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOutNotification/function.json (hunk 13 above)
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/TestScripts/CSharp/NotificationHubOutNotification/function.json]: this used notification type. Earlier were string samples

File: test/WebJobs.Script.Tests/EndToEndTestsBase.cs (hunk 15 above)
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/EndToEndTestsBase.cs]: Will do!

File: test/WebJobs.Script.Tests/TestScripts/Node/NotificationHubOut/function.json (hunk 16 above)
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/TestScripts/Node/NotificationHubOut/function.json]: Right now, no support for auto binding. Will add this later.

File: src/WebJobs.Script/WebJobs.Script.csproj (hunk 11 above)
Comment (pragnagopa) [inline src/WebJobs.Script/WebJobs.Script.csproj]: Yes. Will update these once extension is checked in.

File: sample/NotificationHub-CSharp/function.json [hunk 19]
```diff
@@ -0,0 +1,14 @@
+﻿{
+  "bindings": [
+    {
+      "type": "queueTrigger",
+      "direction": "in",
+      "queueName": "samples-notificationhub-csharp"
+    },
+    {
+      "type": "notificationhub",
```
Comment (mathewc) [inline sample/NotificationHub-CSharp/function.json]: Need to camel case notificationHub properly throughout

File: test/WebJobs.Script.Tests/CSharpEndToEndTests.cs (hunk 6 above)
Comment (fabiocav) [inline test/WebJobs.Script.Tests/CSharpEndToEndTests.cs]: @brettsam currently, if the host is not running, the project.json will **not** be processed. This was a limitation we were aware of with the current design but have plans to address. @mattchenderson had that requirement for something he is working on and we validated that, as long as the host is running, the packages are restored and loaded as expected.
There are some things we can do to simplify the tests so you don't have to worry about the project.json, we can revisit this.

**Comment by brettsam** [comment]
Looks good to me.
